File: alchemist-implementationbase/src/main/java/it/unibo/alchemist/model/implementations/timedistributions/WeibullTime.java
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.implementations.timedistributions;

import it.unibo.alchemist.model.interfaces.Node;
import org.apache.commons.math3.random.RandomGenerator;
import org.apache.commons.math3.util.FastMath;
import org.apache.commons.math3.special.Gamma;
import org.apache.commons.math3.distribution.WeibullDistribution;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import it.unibo.alchemist.model.implementations.times.DoubleTime;
import it.unibo.alchemist.model.interfaces.Environment;
import it.unibo.alchemist.model.interfaces.Time;

/**
 * Weibull distributed events.
 * 
 * @param <T> concentration type
 */
public class WeibullTime<T> extends AbstractDistribution<T> {

    private static final long serialVersionUID = 5216987069271114818L;
    @SuppressFBWarnings(value = "SE_BAD_FIELD", justification = "All the random engines provided by Apache are Serializable")
    private final RandomGenerator rand;
    private final WeibullDistribution dist;
    private final double offset;

    /**
     * @param mean
     *            mean for this distribution
     * @param deviation
     *            standard deviation for this distribution
     * @param random
     *            {@link RandomGenerator} used internally
     */
    public WeibullTime(final double mean, final double deviation, final RandomGenerator random) {
        this(mean, deviation, new DoubleTime(random.nextDouble() * mean), random);
    }

    /**
     * @param mean
     *            mean for this distribution
     * @param deviation
     *            standard deviation for this distribution
     * @param start
     *            initial time
     * @param random
     *            {@link RandomGenerator} used internally
     */
    public WeibullTime(final double mean, final double deviation, final Time start, final RandomGenerator random) {
        this(random, weibullFromMean(mean, deviation, random), 0, start);
    }

    /**
     * @param shapeParameter
     *            shape parameter for this distribution
     * @param scaleParameter
     *            shape parameter for this distribution
     * @param offsetParameter
     *            minimum possible time interval for this distribution
     * @param start
     *            initial time
     * @param random
     *            {@link RandomGenerator} used internally
     */
    public WeibullTime(final double shapeParameter, final double scaleParameter, final double offsetParameter,
            final Time start, final RandomGenerator random) {
        this(random, new WeibullDistribution(random, shapeParameter, scaleParameter,
                WeibullDistribution.DEFAULT_INVERSE_ABSOLUTE_ACCURACY), offsetParameter, start);
    }

    private WeibullTime(final RandomGenerator rand, final WeibullDistribution dist, final double offset, final Time start) {
        super(start);
        this.rand = rand;
        this.dist = dist;
        this.offset = offset;
    }

    @Override
    public final void updateStatus(
        final Time currentTime,
        final boolean executed,
        final double param,
        final Environment<T, ?> environment
    ) {
        if (executed) {
            setNextOccurrence(currentTime.plus(new DoubleTime(genSample())));
        }
    }

    /**
     * @return a sample from the distribution
     */
    protected double genSample() {
        return dist.inverseCumulativeProbability(rand.nextDouble()) + offset;
    }

    /**
     * @return the mean for this distribution.
     */
    public double getMean() {
        return dist.getNumericalMean() + offset;
    }

    /**
     * @return the standard deviation for this distribution.
     */
    public double getDeviation() {
        return FastMath.sqrt(dist.getNumericalVariance());
    }

    @Override
    public final double getRate() {
        return getMean();
    }

    /**
     * Generates a {@link WeibullDistribution} given its mean and standard deviation.
     * 
     * @param mean
     *            the mean
     * @param deviation
     *            the standard deviation
     * @param random
     *            the random generator
     * @return a new {@link WeibullDistribution}
     */
    @SuppressFBWarnings("FL_FLOATS_AS_LOOP_COUNTERS")
    protected static WeibullDistribution weibullFromMean(
        final double mean,
        final double deviation,
        final RandomGenerator random
    ) {
        final double t = FastMath.log((deviation * deviation) / (mean * mean) + 1);
        double kmin = 0, kmax = 1;
        while (Gamma.logGamma(1 + 2 * kmax) - 2 * Gamma.logGamma(1 + kmax) < t) {
            kmin = kmax;
            kmax *= 2;
        }
        double k = (kmin + kmax) / 2;
        while (kmin < k && k < kmax) {
            if (Gamma.logGamma(1 + 2 * k) - 2 * Gamma.logGamma(1 + k) < t) {
                kmin = k;
            } else {
                kmax = k;
            }
            k = (kmin + kmax) / 2;
        }
        final double shapeParameter = 1 / k;
        final double scaleParameter = mean / FastMath.exp(Gamma.logGamma(1 + k));
        return new WeibullDistribution(
            random,
            shapeParameter,
            scaleParameter,
            WeibullDistribution.DEFAULT_INVERSE_ABSOLUTE_ACCURACY
        );
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public WeibullTime<T> cloneOnNewNode(final Node<T> destination, final Time currentTime) {
        return new WeibullTime<>(rand, dist, offset, currentTime);
    }
}

File: alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/timedistributions/WeibullTime.kt
package it.unibo.alchemist.model.timedistributions

import it.unibo.alchemist.model.Environment
import it.unibo.alchemist.model.Node
import it.unibo.alchemist.model.Time
import it.unibo.alchemist.model.times.DoubleTime
import org.apache.commons.math3.distribution.WeibullDistribution
import org.apache.commons.math3.random.RandomGenerator
import org.apache.commons.math3.special.Gamma
import org.apache.commons.math3.util.FastMath

/**
 * Weibull distributed events.
 *
 * @param <T> concentration type
 */
open class WeibullTime<T> private constructor(
    private val randomGenerator: RandomGenerator,
    private val backingDistribution: WeibullDistribution,
    private val offset: Double,
    start: Time,
) : AbstractDistribution<T?>(start) {

    /**
     * @param mean
     *            mean for this distribution
     * @param deviation
     *            standard deviation for this distribution
     * @param random
     *            {@link RandomGenerator} used internally
     */
    constructor(mean: Double, deviation: Double, random: RandomGenerator) : this(
        mean,
        deviation,
        DoubleTime(random.nextDouble() * mean),
        random,
    )

    /**
     * @param mean
     *            mean for this distribution
     * @param deviation
     *            standard deviation for this distribution
     * @param start
     *            initial time
     * @param random
     *            {@link RandomGenerator} used internally
     */
    constructor(mean: Double, deviation: Double, start: Time, random: RandomGenerator) : this(
        random,
        weibullFromMean(mean, deviation, random),
        0.0,
        start,
    )

    /**
     * @param shapeParameter
     *            shape parameter for this distribution
     * @param scaleParameter
     *            shape parameter for this distribution
     * @param offsetParameter
     *            minimum possible time interval for this distribution
     * @param start
     *            initial time
     * @param random
     *            {@link RandomGenerator} used internally
     */
    constructor(
        shapeParameter: Double,
        scaleParameter: Double,
        offsetParameter: Double,
        start: Time,
        random: RandomGenerator,
    ) : this(
        random,
        WeibullDistribution(random, shapeParameter, scaleParameter, PREFERRED_INVERSE_CUMULATIVE_ACCURACY),
        offsetParameter,
        start,
    )

    override fun updateStatus(currentTime: Time, executed: Boolean, param: Double, environment: Environment<T?, *>?) {
        if (executed) {
            this.setNextOccurrence(currentTime.plus(DoubleTime(1.0 / this.genSample())))
        }
    }

    /**
     * @return a sample from the distribution
     */
    protected fun genSample(): Double =
        backingDistribution.inverseCumulativeProbability(randomGenerator.nextDouble()) + this.offset

    /**
     * @return the mean for this distribution.
     */
    val mean: Double
        get() = backingDistribution.numericalMean + this.offset

    /**
     * @return the standard deviation for this distribution.
     */
    val deviation: Double
        get() = FastMath.sqrt(backingDistribution.numericalVariance)

    override fun getRate(): Double = this.mean

    override fun cloneOnNewNode(
        destination: Node<T?>,
        currentTime: Time,
    ): WeibullTime<T?> = WeibullTime(
        this.randomGenerator,
        this.backingDistribution,
        this.offset,
        currentTime,
    )

    protected companion object {
        private const val PREFERRED_INVERSE_CUMULATIVE_ACCURACY = 1.0E-9

        /**
         * Generates a {@link WeibullDistribution} given its mean and standard deviation.
         *
         * @param mean
         *            the mean
         * @param deviation
         *            the standard deviation
         * @param random
         *            the random generator
         * @return a new {@link WeibullDistribution}
         */
        protected fun weibullFromMean(mean: Double, deviation: Double, random: RandomGenerator?): WeibullDistribution {
            val t = FastMath.log(deviation * deviation / (mean * mean) + 1.0)
            var kmin = 0.0

            var kmax: Double
            kmax = 1.0
            while (Gamma.logGamma(1.0 + 2.0 * kmax) - 2.0 * Gamma.logGamma(1.0 + kmax) < t) {
                kmin = kmax
                kmax *= 2.0
            }

            var k: Double
            k = (kmin + kmax) / 2.0
            while (kmin < k && k < kmax) {
                if (Gamma.logGamma(1.0 + 2.0 * k) - 2.0 * Gamma.logGamma(1.0 + k) < t) {
                    kmin = k
                } else {
                    kmax = k
                }
                k = (kmin + kmax) / 2.0
            }

            val shapeParameter = 1.0 / k
            val scaleParameter = mean / FastMath.exp(Gamma.logGamma(1.0 + k))
            return WeibullDistribution(random, shapeParameter, scaleParameter, PREFERRED_INVERSE_CUMULATIVE_ACCURACY)
        }
    }
}
